### PR TITLE
Custom ANTLR-based SQL parser

### PR DESCRIPTION
**Pull request by samredai**
### Summary

This adds an antlr-based custom parser with a more robust AST that can better represent the various query components and their attributes (huge thanks to @CircArgs who did most of the work here!). The AST logic has been added to an `ast2.py` file temporarily as we rework the compile and build. Once we have the new build logic leveraging this AST, we can move it to replace the current `ast.py` file.

The PR also adds a grammar file (largely derived from Spark SQL and Presto's grammar files) and the files generated by the antlr generator. I've added some information to the contributor file on how to use the antlr4 tool to generate the parser based on the contents of the grammar.

### Test Plan

Included tests for round-trip parsing of TPCDS queries and comparing the parse tree at the beginning to the parse tree at the end.

- [x] PR has an associated issue: #271 #393 
- [x] `make check` passes
- [x] `make test` shows 100% unit test coverage

### Deployment Plan

N/A

### CODE REVIEW

**Review by CircArgs — APPROVED** [review]

File: dj/sql/parsing/backends/antlr4.py [hunk 1]
```diff
@@ -0,0 +1,975 @@
...
+    return ast.Lambda(identifiers=identifier, expr=expr)
+
+
+def parse(sql: str) -> ast.Query:
+    """
+    Parse a string into a DJ ast using the ANTLR4 backend.
+    """
+    tree = parse_statement(sql)
+    query = visit(tree)
+    return query
```
Comment (agorajek) [inline dj/sql/parsing/backends/antlr4.py:975]: oprah-emotional

File: tests/sql/parsing/queries/tpcds/test_tpcds.py [hunk 2]
```diff
@@ -1,235 +1,390 @@
...
+# mypy: ignore-errors
+# pylint: skip-file
+from difflib import SequenceMatcher
 
 import pytest
+import sqlparse
+
+from dj.sql.parsing.backends.antlr4 import parse, parse_statement
+
+ansi_tpcds_files = [
```
Comment (agorajek) [inline tests/sql/parsing/queries/tpcds/test_tpcds.py:15]: I can't see these files / variable used anywhere. What am I missing?

**Review by agorajek — APPROVED** [review]
This is huge. Thanks @CircArgs and @samredai for all the work!

File: tests/sql/parsing/queries/tpcds/test_tpcds.py (hunk 2 above)
Comment (samredai) [reply tests/sql/parsing/queries/tpcds/test_tpcds.py:15]: Woah, thanks for catching this! This list just needed to be added to the parametrized test, pushed a commit to update that 👍

**Pull request merged**